certora/specs/Roles.spec

```diff
@@ -1,36 +1,102 @@
+    function multicall(bytes[]) external returns(bytes[]) => NONDET DELETE;
+    function deposit(uint256, address) external returns(uint256) => NONDET DELETE;
+    function mint(uint256, address) external returns(uint256) => NONDET DELETE;
+    function withdraw(uint256, address, address) external returns(uint256) => NONDET DELETE;
+    function redeem(uint256, address, address) external returns(uint256) => NONDET DELETE;
 
-    function _.idToMarketParams(MetaMorphoHarness.Id) external => AUTO;
-    function _.supplyShares(MetaMorphoHarness.Id, address) external => AUTO;
-    function _.accrueInterest(MetaMorphoHarness.MarketParams) external => AUTO;
+    function owner() external returns(address) envfree;
+    function pendingOwner() external returns(address) envfree;
+    function curator() external returns(address) envfree;
+    function guardian() external returns(address) envfree;
+    function isAllocator(address target) external returns(bool) envfree;
 
-    function _.balanceOf(address) external => AUTO;
+    // Sumarize Morpho external calls, as they don't depend on the authorization system of MetaMorpho.
+    function _.idToMarketParams(MetaMorphoHarness.Id) external => CONSTANT;
+    function _.supplyShares(MetaMorphoHarness.Id, address) external => CONSTANT;
+    function _.accrueInterest(MetaMorphoHarness.MarketParams) external => CONSTANT;
+    function _.expectedSupplyAssets(MetaMorphoHarness.MarketParams, address) external => CONSTANT;
+    function _.lastUpdate(MetaMorphoHarness.Id) external => CONSTANT;
+    function _.market(MetaMorphoHarness.Id) external => CONSTANT;
+    function _.supply(MetaMorphoHarness.MarketParams, uint256, uint256, address, bytes) external => CONSTANT;
+    function _.withdraw(MetaMorphoHarness.MarketParams, uint256, uint256, address, address) external => CONSTANT;
 
+    // Summarize MetaMorpho seen as a token, useful for `transferFrom`.
+    function balanceOf(address) internal returns(uint256) => CONSTANT;
+    function allowance(address, address) internal returns(uint256) => CONSTANT;
+    function ERC20._transfer(address, address, uint256) internal => CONSTANT;
+
+    // Summarize other tokens, useful for `skim`.
+    function _.balanceOf(address) external => CONSTANT;
     function SafeERC20.safeTransfer(address, address, uint256) internal => CONSTANT;
-    function SafeERC20.safeTransferFrom(address, address, address, uint256) internal => CONSTANT;
 }
 
+// Check that the owner has more power than the guardian.
+rule ownerIsGuardian(method f, calldataarg args)
+filtered {
+    f -> !f.isView
+}
+{
+    storage initial = lastStorage;
+
+    env e1; env e2;
+    require e1.block.timestamp == e2.block.timestamp;
+    require e1.msg.value == e2.msg.value;
+
+    require e1.msg.sender != 0;
+    require e2.msg.sender != 0;
+
+    require e1.msg.sender == guardian();
+    require e1.msg.sender != owner();
+    require e1.msg.sender != pendingOwner();
+    require e1.msg.sender != currentContract;
```

Comment:
Good point, it's not needed anymore. It was needed before when I was trying to also include the main functions (deposit, mint, withdraw, redeem), because the current contract (theoretically) `transfer` it's own funds